examples/benchmark/datagen.py

```diff
@@ -0,0 +1,354 @@
+    serializing call to the `next` method of given iterator/generator.
+    """
+
+    def __init__(self, it):
+        self._it = it
+        self._lock = Lock()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        with self._lock:
+            return self._it.__next__()
+
+
+def build_pipeline(
+    num_samples: float, resolution: Tuple[int, int], shard_size: float
+) -> Iterator[LabelledSample]:
+    # define the pipeline to generate the dataset
+    # TODO: parameterize the data generator to allow for creating other kinds of datasets(eg: text).
+    sample_generator = ImageGenerator(*resolution)
+    pipeline: Iterator[LabelledSample] = sample_generator.create(
+        iter(range(int(num_samples)))
+    )
+    if shard_size:
+        monotonic_ctr = ThreadSafeCounter()
+        pipeline: Iterator[List[LabelledSample]] = Utils.batcher(
+            items=pipeline,
+            size_extractor=lambda item: item.data.getbuffer().nbytes,
+            size_threshold=shard_size,
+        )
+        pipeline: Iterator[LabelledSample] = (
+            Utils.tar_samples(batch, monotonic_ctr) for batch in pipeline
+        )
+
+    return ThreadSafeIterator(pipeline)
+
+
+def build_producers(
+    num_workers: int, queue: Queue, dataset_generator: Iterator[LabelledSample]
+) -> List[Thread]:
+    barrier = Barrier(num_workers)
+    return [
+        Thread(
+            target=producer,
+            name=f"DataGenerator-{i}",
+            kwargs={
+                "identifier": i,
+                "queue": queue,
+                "generator": dataset_generator,
+                "barrier": barrier,
+            },
+        )
+        for i in range(num_workers)
+    ]
+
+
+def build_consumers(
+    num_workers: int, queue: Queue, disambiguator: str, region: str, s3_bucket: str
+) -> List[Thread]:
+    return [
+        Thread(
+            target=consumer,
+            name=f"Uploader-{i}",
+            kwargs={
+                "identifier": i,
+                "queue": queue,
+                "activity": lambda sample: Utils.upload_to_s3(
+                    region=region,
+                    data=sample.data,
+                    bucket=s3_bucket,
+                    key=f"{disambiguator}/{sample.label}",
+                ),
+            },
+        )
+        for i in range(num_workers)
+    ]
+
+
+def producer(generator: Iterator, barrier: Barrier, queue: Queue, identifier: int):
+    for item in generator:
+        queue.put(item)
+    # wait for all producers to finish
+    barrier.wait()
+    # signal that there are no further items
+    if identifier == 0:
+        queue.put(Sentinel())
+
+
+def consumer(queue: Queue, activity: Callable[[LabelledSample], None], identifier: int):
+    while True:
+        # click.echo(f"Consumer running on thread {threading.current_thread().ident}")
+        item: Union[LabelledSample, Sentinel] = queue.get()
+        if type(item) is Sentinel:
```

Comment:
Nit: Is there any reason we couldn't put in the Sentinel class itself in, and then just `if item is Sentinel`?